### ISSUE TITLE
Need of x11

### ISSUE THREAD

**Issue by kinow** [BSC-ES/autosubmit#544]
In GitLab by @orioltinto on Jun 12, 2020, 12:42

In a workflow I'm working on we have some jobs that require to have a display defined and to have -x11 enabled even there's anything for the user to do interactively.

A simple test to reproduce the problem is to execute a simple template with:
`xclock`

I think that it should be possible to do it through paramiko.

**Comment by kinow** [comment]
In GitLab by @mcastril on Jun 12, 2020, 13:04

I could launch x11 inside a batch job with this headers:

```
#SBATCH --qos=bsc_es
#SBATCH -A bsc32
#
#
#
#SBATCH --cpus-per-task=1
#SBATCH -n 48
#SBATCH -t 00:05:00
#SBATCH -J a2wz_REMOTE_SETUP
#SBATCH --output=/gpfs/scratch/bsc32/bsc32252/a2wz/LOG_a2wz/a2wz_REMOTE_SETUP.20200612123431.out
#SBATCH --error=/gpfs/scratch/bsc32/bsc32252/a2wz/LOG_a2wz/a2wz_REMOTE_SETUP.20200612123431.err
#SBATCH --x11=batch
```

The important one for this case is `--x11=batch`

This was run after checking that I could run `xclock` from the login terminal. The problem, as @orioltinto accurately pointed out, it seems it is that Paramiko is not correctly forwarding X11.

I saw some messages regarding this matter, so maybe it is possible:

https://stackoverflow.com/questions/12354047/x11-forwarding-with-paramiko

@dbeltrankyl @wuruchi

**Comment by kinow** [comment]
In GitLab by @mcastril on Jun 12, 2020, 13:32

If you want to test you can copy my experiment a2wz. The template is in the same experiment (I used local, creating the directory in the experiment itself).

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jul 23, 2020, 10:34

Hi

I've worked on this but so far I didn't get satisfying results.

Paramiko is not able to use x11 forwarding at ssh client level ( at least I didn't found it), instead it has to create a new session on the transport layer and in this layer, it can request an x11 connection. 
This x11 connection is not implemented on paramiko itself it has to be added manually through a handler using the Xlib library, so essentially paramiko only adds the possibility of establishing a connection.

After adding manually a handler for x11 (using xlib), you can use simple commands like exec_command("xclock") however it only will work though direct commands like this one since the connection will be closed once the command finishes since it is done on a session-level( alias one command).

Due to the nature of job submission, the issue is that paramiko can't know that there will be an x11 request somewhere in the future when the job begins.  

So the main issue resides on that paramiko can't forward ssh connections of slurm batch since the session will be closed

So in summary, the main issue resides on the way that autosubmit launches a job though paramiko. It creates a connection with exec-command and this connection only includes the submission of the jobs(multiple at once) not the execution of the command.

So the solution could be that paramiko perma-open an x11 channel and receive all x11 connections from the cluster though a thread on autosubmit, however as I said, I don't see that this possibility is implemented on paramiko library.

 When @mcastril returns from holidays we will discuss maybe there are something that I didn't find.

Edit: Also this connection is always established from paramiko to the cluster, there is no way to "wait for connection" from cluster to paramiko.

**Comment by kinow** [comment]
In GitLab by @mcastril on Aug 10, 2020, 12:58

In the meeting last week we discussed two possibilities:

* Try using the interactive partition. An example is the COMPILATION step in Auto-EC-Earth workflow.
* Try declaring a PS machine, which runs things directly in the login. This is not the best solution but at least will help to know if it is possible to achieve an X11 connection through Autosubmit.

**Comment by kinow** [comment]
In GitLab by @mcastril on Oct 1, 2020, 13:47

Oriol, did you look if it was working using -p interactive? I don't remember if we did the test.

**Comment by kinow** [comment]
In GitLab by @mcastril on Oct 15, 2020, 17:25

mentioned in issue ces/nemo_performance_profile#41

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on May 13, 2021, 13:48

Hello @mcastril , 

I have a working demo, with a very reduced version of autosubmit ( only submit a template get jobid and queue status). I only tried to display xclock and a xterm. 

I tried two schemes

* Using SRUN, since it is an interactive command thus paramiko won't close the channel until the srun script finishes, it works in paramiko adapting a bit the work done previously, however, there is a big downside for use srun within autosubmit since I would need to change so many things for run autosubmit with srun. 

* With sbatch, since it is not interactive, paramiko just close the channel when it gets the EOF of the job submitted. However, I did a bit of hack and launched the command with a (sbatch -D job_script && sleep infinity) this keeps the channel open forever and thus the X11 connections can be redirected to our local display successfully.

The sbatch solution has the issue that the channel is blocking, but with threads and with a SACCT command ( if job is completed => close the channel) I can close the channel manually and avoid the block.

I will implement the sbatch solution demo into autosubmit code.

For now, I adapted the work done with x11 to 3.14 version, and completed the demo with threads. So it is matter of time

**Comment by kinow** [comment]
In GitLab by @mcastril on May 13, 2021, 13:53

Congratulations @dbeltrankyl! I think it's a very smart solution. It's clear that there was not an easy way to deal with this. As it will only be used in very specific situations (such as launching Paraver) I would minimize the impact on Autosubmit code, so I think that option 2 is the wisest as you suggest.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jul 29, 2021, 14:43

Hello, 

I tried to use paraver within this X11 implementation. 

* In the code isolated that I had in a demo. 
* Within Autosubmit. 

The job runs without issues but the graphical screen doesn't show up as it does with xclock. 

I think that perhaps there is an issue triggered when using EXPORT=module load paraver; since it is added into the command itself. 

I will test with other configurations as soon as possible

**Comment by kinow** [comment]
In GitLab by @rocsalvadorbsc on Aug 3, 2023, 12:31

Hi,

I know this issue was stopped a long ago, but now since I am porting the automatic performance profiling tool to autosubmit and it uses Paraver, I need to use X11. I don't know if there's any possibility for it to work in Marenostrum 4 or if I should wait for Marenostrum 5.

Thank you

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 3, 2023, 14:32

Seems that works for local but... for mn4 I have bad news


I just remind this mail from 21/06/22

```
Dear users,

Slurm has been successfully upgraded on MN4.

Command sbatch does not support x11 directive, please, use salloc or srun instead.

In case any issue arises, please, do not hesitate to contact us.

Thanks for your understanding.

BSC Support Team.
```

This is a big issue because the actual code relies upon using --x11=batch .. all jobs are submitted using sbatch not sure if a workaround is possible, FYI @mcastril

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 3, 2023, 14:33

Also, not sure if marenostrum5 will have the same issue

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 3, 2023, 14:55

I guess I should prompt a salloc instead of sbatch when dealing with X11

current issues without thinking much:

* salloc doesn't understand #sbatch directives -> Need to make an auto translator into command line options for all the directives on the cmd... 
* salloc is interactive -> Need to encapsulate it into a thread 
  this is tricky.. as Autosubmit expects the sbatch behavior, luckily it returns a jobid instantaneous so it seems doable

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 3, 2023, 16:09

Miguel told me that he will ask to David Vicente

**Comment by kinow** [comment]
In GitLab by @rocsalvadorbsc on Aug 8, 2023, 10:49

mentioned in issue ces/automatic_performance_profiling#41

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 10, 2023, 12:07

I tried and made some progress in the salloc implementation. Still, it is not working so far, multiple issues are raised across Autosubmit code, and paramiko library is not getting along with that command.
 
There is little to nothing info on the internet about salloc + paramiko + x11 implementation, and the pycharm debugger is not working with x11 enabled, so the task is pretty time-consuming. 

I tried to make it work before my holidays, but I think I won't do it in time ( starting the following Monday until 28th ).

I need to work on another issue, but If I have time left, I'll keep trying

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 29, 2023, 14:41

Hello @rocsalvadorbsc , 

X11 is working with wxparaver. I did a straightforward script.sh

`module load BSCTOOLS; wxparaver`

And it shows up on my home computer. 

I'm uploading a temporal module based on the x11 branch in the BSC machine because one issue is still pending ( sometimes, the log is not working for these jobs ).

To load this version, you have to perform.

`module load autosubmit/4.0.88-foss-2015a-Python-3.7.3`

To activate x11 in a job, you have to add these parameters in your selected job:

```yaml
JOBS:
  X11_JOBS:
    FILE: "your_script"
    WALLCLOCK: #must match --time
    X11: True
    X11_OPTIONS: "-n 4 --qos=debug --partition=debug --x11=all --time=00:30:00" 
    #example

```

NOTE: X11 jobs will ignore the standard settings because it is a different command, so you must put x11_option. This will be improved, but for now this works

**Comment by kinow** [comment]
In GitLab by @rocsalvadorbsc on Aug 29, 2023, 16:00

Hi @dbeltrankyl,

I tried to run the paraver job defined at `/esarchive/autosubmit/a68z/proj/auto-performance-profiling/autosubmit/configuration/common/jobs.yml`:
```
  paraver:
    FILE: autosubmit/templates/paraver.sh
    WALLCLOCK: 02:00  
    X11: True
    X11_OPTIONS: "-n 4 --qos=debug --partition=debug --x11=all --time=02:00:00" 
    RUNNING: member
```
But I got the following error (it didn't generate an error log): 

`Invalid MIT-MAGIC-COOKIE-1 keyError: Unable to initialize gtk, is DISPLAY set properly?`

I am running it from the workstation.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 30, 2023, 08:35

Hello @rocsalvadorbsc ,

Are you connected to the AS machine with ssh -XY?

Can you open a terminal xclock or gedit or any GUI app? to ensure that it is not an issue with the module

**Comment by kinow** [comment]
In GitLab by @rocsalvadorbsc on Aug 30, 2023, 08:58

Hi @dbeltrankyl,

I was running it directly from the workstation with no ssh. Now, I tried to run it with `ssh -XY` from my laptop (gedit opens with no issue) but I get this error when the script reaches the paraver command:

`X11 connection rejected because of wrong authentication.`

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 30, 2023, 10:53

Hello @rocsalvadorbsc ,

I did an update to solve the debug messages and timeout in psplatform that was higher than intended. 

Two things: 

* When using TYPE: PS, autosubmit is not using salloc. Instead, it will use the login node directly; thus, the connection will be synchronous. If the job crashes, autosubmit will be stuck until the timeout is reached, so you will need to put a lower wallclock in that case. To avoid this, you can use a platform with TYPE: slurm, and Autosubmit will use salloc. ( I recommend doing in that way )

* It is working on my home computer, with PS and SLURM. I have the same issue that you have when using the office machines. I'm not sure why. I am still debugging it.

FYI @albertvilabsc, sorry for disturbing you, but do you know why this is happening? For summarized context:

I've added an X11 option in autosubmit. It is working completely fine on the BSC laptop, but when I try to run the same code in the office or autosubmit the VM, the job complains about 

`X11 connection rejected because of wrong authentication.`

I tried to do this https://unix.stackexchange.com/questions/674214/x11-connection-rejected-because-of-wrong-authentication, but it is not working.

There is no difference in the target machine ( marenostrum0 in this case ), so I guess the issue is in some configuration of the office machines. What do you think?

**Comment by kinow** [comment]
In GitLab by @albertvilabsc on Aug 30, 2023, 11:48

Hi @dbeltrankyl ,

Can you send me the intructions to reproduce this in the WS and autosubmit VM?

Many thanks,

Albert

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 30, 2023, 13:21

Hello @albertvilabsc ,

Sorry for the delay. I was performing more unsuccessful tests. 

You perform:

* module load autosubmit/4.0.88-foss-2015a-Python-3.7.3
* autosubmit expid -y a6en -H "marenostrum0" -d "x11-test" # it will give you an expid
* modify /esarchive/autosubmit/$expid/conf/platforms.yml and change the USER for your own
```yaml
  marenostrum0:
    TYPE: ps
    HOST: mn0.bsc.es
    PROJECT: bsc32
    USER: bsc32070 # <---------
    SCRATCH_DIR: /gpfs/scratch
    MAX_WALLCLOCK: 48:00
    MAX_PROCESSORS: '2400'
    PROCESSORS_PER_NODE: '48'
    SERIAL_QUEUE: debug
    QUEUE: debug
```

Once this is done, to run one or more time is only required to do the following:

* autosubmit create $expid -np
* autosubmit run $expid

The executed script can be found and modified in `/esarchive/autosubmit/$expid/proj/test_x11/paraver_job.sh`
 

The job options can be found in 
`/esarchive/autosubmit/$expid/conf/jobs_$expid.yml`

**Comment by kinow** [comment]
In GitLab by @albertvilabsc on Aug 30, 2023, 13:26

@dbeltrankyl I will make some tests soon. In the bsceshub it is working?

Thanks,

Albert

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 30, 2023, 13:41

I did not test it in the bsceshub only VM and office ones (and in my laptop in which it is working ) but I can upload it and test it

**Comment by kinow** [comment]
In GitLab by @albertvilabsc on Aug 30, 2023, 13:48

Yes please, if you can test in the bsceshub, will be perfect!

Thanks,

Albert

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 30, 2023, 15:18

Hello,

I've uploaded it to bschubs 

In case of permission issues in my /home/Earth/dbeltran, perform 

`cp -r /esarchive/autosubmit/a6en/proj /esarchive/autosubmit/$expid/.`

Before the create.


I've tested it, and the same issue is rising.

```bash
timeout 156.0 nohup bash /gpfs/scratch/bsc32/bsc32070/a6en/LOG_a6en/a6en_PARAVER.cmd > /gpfs/scratch/bsc32/bsc32070/a6en/LOG_a6en/a6en_PARAVER.cmd.out.0 2> /gpfs/scratch/bsc32/bsc32070/a6en/LOG_a6en/a6en_PARAVER.cmd.err.0 & echo $! ; sleep 156.0 2>/dev/null
X11 connection rejected because of wrong authentication.
```

**Comment by kinow** [comment]
In GitLab by @mcastril on Aug 30, 2023, 21:00

I guess that you already checked the $DISPLAY variable and etc.

We had a related discussion in the Docker MR:

https://earth.bsc.es/gitlab/es/autosubmit/-/merge_requests/364#note_227693

In the discussion, Bruno was pointing to this interesting [post](https://goteleport.com/blog/x11-forwarding/)

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 31, 2023, 08:12

> In the discussion, Bruno was pointing to this interesting [post](https://goteleport.com/blog/x11-forwarding/)

Thanks, Miguel I'll look into it.

> I guess that you already checked the $DISPLAY variable and etc.

Autosubmit should be able to do it automatically, like on my laptop, but just in case, I manually exported these parameters, and it did not work.

**Comment by kinow** [comment]
In GitLab by @albertvilabsc on Aug 31, 2023, 08:23

The only difference in the BSC laptop is that there I force to use x11 instead of Wayland.

Link: https://linuxconfig.org/how-to-disable-wayland-and-enable-xorg-display-server-on-ubuntu-18-04-bionic-beaver-linux

I will take a look too.

Thanks,

Albert

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 31, 2023, 09:42

Thanks, Albert, for looking at it. 

I can discard a network/vpn issue because it still works under the bsc network on my laptop.

I compared the job printenv and found no significant differences...


```bash
< XDG_SESSION_ID=54866
---
> XDG_SESSION_ID=54845

< SSH_CLIENT=84.88.184.211 44276 22
---
> SSH_CLIENT=84.88.185.220 53828 22

< SSH_CONNECTION=84.88.184.211 44276 84.88.52.168 22
---
> SSH_CONNECTION=84.88.185.220 53828 84.88.52.168 22
< DISPLAY=localhost:18.0
---
> DISPLAY=localhost:16.0
```

**Comment by kinow** [comment]
In GitLab by @albertvilabsc on Aug 31, 2023, 09:53

Thanks to test it!

Albert

**Comment by kinow** [comment]
In GitLab by @rocsalvadorbsc on May 16, 2024, 10:01

Hi @dbeltrankyl,

has there been any progress on this issue?

Thank you

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on May 16, 2024, 10:49

The last code was working under my laptop and using Marenostrum4, in hubs never worked

I'll test when I have a bit of time if we're lucky with Marenostrum5

#1246

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on May 17, 2024, 12:55

Hello @rocsalvadorbsc ,

I'm having the same issue, in laptop works after the fixes to make it compatible to mn5

But In the hubs I still get:

```
X11 connection rejected because of wrong authentication.
```

You could try to deploy the  v4.1.7-x11-test branch on your laptop and use it to handle these x11 jobs. However, you won't be in esarchive. 

Maybe you can install that autosubmit version on your laptop and mount esarchive and point your laptop /etc/autosubmitrc to the esarchive.DB, but I'm not sure if it will work properly as I never tried to do it.

If you want to try this approach let me know because I can upload a pip version with these changes which would be easier to deploy

Edit: The laptop approach (at least with a locally isolated autosubmit.db ) should work if you use the bsc one .. because it would be the same system as mine.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on May 17, 2024, 13:01

To use X11 I configured the job as: 

```yaml
JOBS:
  SIM:
    FILE: SIM.sh
    RUNNING: chunk
    WALLCLOCK: 00:05
    X11: True
    X11_OPTIONS: "-A bsc32 --qos=gp_debug --x11=all --time=02:00:00"

PLATFORMS:
  MARENOSTRUM5:
    TYPE: slurm
    HOST: mn5-gpp2
    PROJECT: bsc32
    USER: bsc032070
    QUEUE: gp_debug
    SCRATCH_DIR: /gpfs/scratch
    ADD_PROJECT_TO_HOST: false
    MAX_WALLCLOCK: 48:00
    TEMP_DIR: ''

.ssh/config
Host mn5-gpp2
    User bsc032070
    HostName glogin2.bsc.es
    ForwardAgent yes
    IdentityFile /home/Earth/dbeltran/.ssh/id_ed25519
```

Where sim.sh contains a single `xclock`

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jun 5, 2024, 10:57

Hello all, 

@rocsalvadorbsc  has been testing the 4.1.7-x11 branch, and after some fixes we managed to get paraver working, 

First we tried with an xclock in which he had this issue: `Invalid MIT-MAGIC-COOKIE-1` which was solved by executing `xhost +local`.

Second, when using the actual template having paraver, autosubmit was not able to check the status of the job. After some fixing in AS code now it is working. 

Third, for invalid salloc options, autosubmit was not complaining and the job was never submitted

I also fixed an issue with the logs, we are not able to redirect salloc jobs logs so I disable the log retrieval for the x11 as they're written in the nohup file and in the aslogs/run_command log.

So, now we confirmed that the X11 is working in two different laptops and with real code, but in the hubs is not working. 

The branch will be merged to master and released on 4.1.7

**Comment by kinow** [comment]
In GitLab by @xyepes on Jun 5, 2024, 11:19

Thanks a lot @dbeltrankyl! Do you think it can be solved the issue in the hubs?

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jun 5, 2024, 11:24

I tried to solve it, but I couldn't https://earth.bsc.es/gitlab/es/autosubmit/-/issues/544#note_229328

@albertvilabsc said that:

> The only difference in the BSC laptop is that there I force to use x11 instead of Wayland.

Maybe that's the issue, we can try again once the version is released

**Comment by kinow** [comment]
In GitLab by @rocsalvadorbsc on Jun 6, 2024, 08:49

mentioned in issue ces/hpc-for-es-team/automatic_performance_profiling#76

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jun 11, 2024, 11:22

4.1.7 is now available in the hubs, but I haven't had time to test it yet. I created a new issue, https://earth.bsc.es/gitlab/es/autosubmit/-/issues/1321, to continue the tests/ discussion.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jun 11, 2024, 11:22

Closing this issue as the X11 code is implemented in 4.1.7, for bschubs we can continue in #1321
